### PR TITLE
[Cmake] Several fixes for TARGET usage

### PR DESCRIPTION
**Pull request by fuzzard**
## Description
Increase support for TARGET usage in the endeavour to continue modernising some of our cmake usage

## Motivation and context
Garbear came across an interesting failure with homebrew and the recent tinyxml2 addition. Dug into it a bit, and it highlighted some shortcomings of TARGET usage in our current macro system.

@wsnipex if you have some time, no real drama's if you dont

## How has this been tested?
Build macos aarch64. In particular built with and without existing ffmpeg

## What is the effect on users?
N/A

## Types of change

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by neo1973** [comment]
@fuzzard: What do you think about making the Wayland protocols a real library like here: https://github.com/neo1973/xbmc/commit/dc9aad7f03b25e0ba7f6cd1f16875c48267c644e

Maybe even the whole Wayland stuff from ExtraTargets.cmake can then be moved to wayland.cmake?

**Comment by notspiff** [comment]
Object library seems more appropriate if it is a single source.

lør. 12. aug. 2023, 15:07 skrev Markus Härer ***@***.***>:

**Comment by fuzzard** [comment]
Definitely something that can be looked at.

**Comment by neo1973** [comment]
I cleaned up my Wayland change a little bit and removed some webOS duplication: https://github.com/neo1973/xbmc/commit/536ba3a44113521cc6fb7ef14181507661176624

Feel free to cherry-pick if you like to.

> Object library seems more appropriate if it is a single source.

Unfortunately I didn't manage to make this work, the problem seems to be here:
```cmake
set(PLATFORM_GLOBAL_TARGET_DEPS $<TARGET_OBJECTS:wayland-extra-protocols>)
```

File: cmake/modules/FindFlatBuffers.cmake [hunk 1]
```diff
@@ -46,7 +46,7 @@ if(FLATBUFFERS_FOUND)
   set(FLATBUFFERS_INCLUDE_DIRS ${FLATBUFFERS_INCLUDE_DIR} ${FLATBUFFERS_MESSAGES_INCLUDE_DIR})
 
   if(NOT TARGET flatbuffers::flatbuffers)
-    add_library(flatbuffers::flatbuffers UNKNOWN IMPORTED)
+    add_library(flatbuffers::flatbuffers INTERFACE IMPORTED)
```
Comment (garbear) [inline cmake/modules/FindFlatBuffers.cmake:49]: What's the difference between these two?

File: cmake/modules/FindCrossGUID.cmake [hunk 2]
```diff
@@ -92,12 +92,12 @@ if(CROSSGUID_FOUND)
     if(CROSSGUID_LIBRARY_RELEASE)
       set_target_properties(CrossGUID::CrossGUID PROPERTIES
                                                  IMPORTED_CONFIGURATIONS RELEASE
-                                                 IMPORTED_LOCATION "${CROSSGUID_LIBRARY_RELEASE}")
+                                                 IMPORTED_LOCATION_RELEASE "${CROSSGUID_LIBRARY_RELEASE}")
```
Comment (garbear) [inline cmake/modules/FindCrossGUID.cmake:95]: A similar change is missing in FindTinyXML.cmake, FindMySqlClient.cmake and FindMariaDBClient.cmake.

Comment (fuzzard) [reply cmake/modules/FindCrossGUID.cmake:95]: Its not required in them as we arent actually using the target created. Its still just using the *_LIBRARIES/*_INCLUDEDIR etc

File: cmake/modules/FindFlatBuffers.cmake (hunk 1 above)
Comment (fuzzard) [reply cmake/modules/FindFlatBuffers.cmake:49]: an unknown imported target is done when you dont know if the lib is static/shared/executable. these states are intended to have a file on disk. An interface doesnt require a file to be physically present, but allows to propagate other properties of targets when linked.

File: cmake/modules/FindRapidJSON.cmake [hunk 3]
```diff
@@ -34,8 +34,6 @@ if(ENABLE_INTERNAL_RapidJSON)
 
   set(RapidJSON_INCLUDE_DIR ${${MODULE}_INCLUDE_DIR})
```
Comment (garbear) [inline cmake/modules/FindRapidJSON.cmake:35]: Not touched in this PR, but doesn't this line have no effect?

File: cmake/modules/FindFlatBuffers.cmake (hunk 1 above)
Comment (garbear) [reply cmake/modules/FindFlatBuffers.cmake:49]: Got, so `UNKNOWN` lib/exe on disk, `INTERFACE` instead to propagate properties

Comment (fuzzard) [reply cmake/modules/FindFlatBuffers.cmake:49]: not exactly. https://cmake.org/cmake/help/latest/command/add_library.html#imported-libraries for some info

**Comment by garbear** [comment]
Not sure if this is related, but this branch failed to build on Ubuntu. Error was:

```
[100%] Linking CXX executable kodi.bin
ld.lld: error: unable to find library -lgenerate-wayland-extra-protocols
collect2: error: ld returned 1 exit status
```

Skipping this PR allowed the build to succeed.

**Comment by fuzzard** [comment]
I'm aware. It's WIP for a reason

File: cmake/modules/FindSpdlog.cmake [hunk 4]
```diff
@@ -2,34 +2,41 @@
...
+  if(TARGET fmt::fmt)
+    # Check if we want to force a build due to a dependency rebuild
+    get_property(LIB_FORCE_REBUILD TARGET fmt::fmt PROPERTY LIB_BUILD)
+  endif()
 
   set(MODULE_LC spdlog)
   SETUP_BUILD_VARS()
 
+  # Darwin systems we want to avoid system packages. We are entirely self sufficient
+  # Avoids homebrew populating rubbish we cant control
```
Comment (garbear) [inline cmake/modules/FindSpdlog.cmake:28]: I recommend the leading prepositions: "On Darwin systems..."

And I'll be trying this on my 2013 macbook. 

```cmake
  # On Darwin systems we want to avoid system packages. We are entirely self
  # sufficient. Avoids homebrew populating rubbish we cant control.
  if(CMAKE_SYSTEM_NAME STREQUAL "Darwin")
    set(_spdlog_find_option NO_SYSTEM_ENVIRONMENT_PATH)
  endif()
```

**Comment by garbear** [comment]
Reading `FindSpdlog.cmake`, I see a lot of checks if the target exists, should we wrap the whole file in `if(NOT TARGET`?

**Comment by fuzzard** [comment]
Jenkins build this please

**Comment by fuzzard** [comment]
So this led into an insane amount of code churn that i wasnt expecting initially.
Im not actually sure how to make this bisectable, as the full conversion of TARGET usage done requires the last two commits to be functional, but without them everything breaks.

The first half dozen or so are ok, but everything else probably needs to be merged together. Would be nice to not have that, but its all intertwined, and if you do one lib, you gotta do them all.

@wsnipex if you find time. Anyone else, if you think something needs clarifying/docs for the average dev to understand things, let me know.

The outcome of this essentially brings all the cmake only packages to full target usage that allows detection of existing built libs to either build if version doesnt match, or just imports the targets. Some minor things around libdvd* probably need further work, but its best left as is for now.
The one caveat that all this has is for multi config generators (VS/Xcode). Theres always been a shortcoming around detection of one config type, and only adding a build dependency for the types not detected (eg debug lib exists, release does not). In this instance, we would want to build the lib only if the release config type is being made, but cmakes generator expressions dont give us a way to do this (that i have found currently).

What this means, if you build a Debug version of kodi, it populates a debug lib into the depends path. If you clear the build folder, and regenerate, it will find the debug lib, and wont add the internal build as one of the lib variants was available. This may cause a failure to build Release variant if it has a postfix debug variant at a worst case, or it builds and has the debug version linked of that lib (less optimised).
The aim is to reduce rebuilding if not needed. In general windows users who use BuildSetup.bat only build in release, so they wont get affected, but devs who build different config variants may need to clear out the depends paths to get a full clean build.

The above isnt new to this PR. its an issue ive been tackling with cmake for a long time now, causing the continued migration to be slowed down immensely. Whether we continue bring stuff internally or not really depends on how much hassle that brings to people. We could always just build the target always, but trying to reduce the old "I dont want to rebuild, i want libs prebuilt" push back from windows devs brings us to this.

**Comment by garbear** [comment]
It seems like the solution here inherits some problems from before. I'd say don't worry about fixing the world, and stick your explanation about the current limitations somewhere in the code or docs.

**Review by wsnipex — APPROVED** [review]
I admit I didn't test this, but it does look good to me.
Let's get this in so we can move forward and deal with possible corner cases when they come up

**Comment by howie-f** [comment]
@fuzzard i'm very sorry :)
```
CMake Error at cmake/modules/FindFFMPEG.cmake:128 (set_target_properties):
  INTERFACE_LIBRARY targets may only have whitelisted properties.  The
  property "FOLDER" is not allowed.
Call Stack (most recent call first):
  cmake/modules/FindFFMPEG.cmake:387 (buildFFMPEG)
  cmake/scripts/common/Macros.cmake:376 (find_package)
  cmake/scripts/common/Macros.cmake:390 (find_package_with_ver)
  CMakeLists.txt:226 (core_require_dep)
```

cmake version 3.18.4 (bullseye). i didn't want to use it anymore

**Comment by mglae** [comment]
This commit is breaking LE12 Kodi build:  [192.log](https://github.com/xbmc/xbmc/files/12567972/192.log)

Error is:
```
CMake Error in xbmc/cores/RetroPlayer/process/X11/CMakeLists.txt:
  Imported target "LibDvdNav::LibDvdNav" includes non-existent path

    "/home/docker/LibreELEC.tv/build.LibreELEC-x11.x86_64-12.0-devel-mg/build/kodi-efe703a063e10b8bbe0493c0c88b75805364c875/depends/include"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.
```

Any recommendations?

**Comment by fuzzard** [comment]
No idea offhand. guess a bad path is being set for LIBDVDNAV_INCLUDE_DIR somehow/somewhere.

**Comment by fuzzard** [comment]
@howie-f your issue should be fixed in https://github.com/xbmc/xbmc/pull/23746

**Comment by mglae** [comment]
@fuzzard LE build is fixed in the meantime: LibreELEC/LibreELEC.tv#8130. Good guess.

**Comment by fuzzard** [comment]
Good to hear. I did look at the FindLibDVD* modules, and did notice we dont really easily accomodate system libs. Im assuming theres a reason, but as enen works on the upstream libdvd stuff, maybe we'll get to a point to just treat them as any normal system provided lib instead of building always.

**Comment by enen92** [comment]
> Good to hear. I did look at the FindLibDVD* modules, and did notice we dont really easily accomodate system libs. Im assuming theres a reason, but as enen works on the upstream libdvd stuff, maybe we'll get to a point to just treat them as any normal system provided lib instead of building always.

Well we've just celebrated the 1 year vfs patch anniversary...so I wouldn't put my hand on the fire for it. Looks pretty much unmaintained (or the videolan focus is elsewhere)
